### ISSUE TITLE
godot buttons can't be resized to 8×8

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot#35005]
**Godot version:**
3.1.2

**OS/device including version:**
windows 10 64 bit

**Issue description:**
I can't shrink buttons below the size 8x16.848
![image](https://user-images.githubusercontent.com/2453707/72199213-f510b500-33ed-11ea-8025-ddabdb331412.png)

**Steps to reproduce:**
- 1 make 24x24 pixel square for the button panel style
- 2 apply the style to all styles in the button except focus use empty style for that.
- 3 enable clip text and clip content
- 4 set the minimum size to 8x8
- 5 try to set the size to 8x8.

**Minimal reproduction project:**
[PocketQuest.zip](https://github.com/godotengine/godot/files/4049012/PocketQuest.zip)

please fix this every time I come back to godot ui I find this is still here.
content should always be resizeable unless the user themselves set a minimum size.

things that should also be looked at..
- anything that restricts the size of the control should be turned off by default
- allowing the user to resize the controls in containers

**Comment by groud** [comment]
This is an expected behavior, the buttons get their minimum size according to their defined style. If you need to resize a button to a very little value like 8x8, you should create a stylebox that have little to no margins.

Regarding your proposals at the end, you should open a detailed proposal in the godot-proposals repository. As I already told you it is unlikely to be accepted, but you can still try to see what the community thinks about it.

**Comment by timothyqiu** [comment]
It's strange that I can't reproduce the issue on macOS. The button in the reproduction project can be resized to 8x8 successfully in 3.1.2 stable.

![demo](https://user-images.githubusercontent.com/372476/72200410-23d96a80-3484-11ea-9fed-681d7611e708.gif)

**Comment by groud** [comment]
Hmm, that diserves more testing then. It not expected that the two behave differently.

**Comment by Shadowblitz16** [comment]
@groud setting the style should override the theme.
there is no point in obeying the theme if it doesn't use it.

btw why the fuck does the theme override the size settings anyways?
it should set the size when using it not in the theme itself

**Comment by Calinou** [comment]
@Shadowblitz16 I have to remind you that we have a [Code of Conduct](https://godotengine.org/code-of-conduct). Please stay constructive.

**Comment by timothyqiu** [comment]
FYI: I just tried on a Windows 10 64 bit machine, and I can't reproduce this either.

The button in the reproduction project can be resized to 8x8 successfully in 3.1.2 stable.

![demo](https://user-images.githubusercontent.com/372476/72252610-8ecca200-363a-11ea-99c6-ea8cbcde5d9c.gif)

**Comment by Shadowblitz16** [comment]
try setting the font to [empty] and adding a icon to the button itself.
both seem to effect the minimum size of the button despite having clip text and clip content on

**Comment by rakkarage** [comment]
![NinePatchMargin](https://user-images.githubusercontent.com/962340/82237172-40d6ec00-9903-11ea-8b51-9c36066d0b30.gif)

Here is another example of the same problem i think?

![Button](https://user-images.githubusercontent.com/962340/82237371-84c9f100-9903-11ea-9d9f-5b56a7355549.png)
![ButtonFocus](https://user-images.githubusercontent.com/962340/82237372-84c9f100-9903-11ea-8378-fa54cd2f0d5e.png)
![ButtonHover](https://user-images.githubusercontent.com/962340/82237374-85628780-9903-11ea-90e7-e2dbe44e9b47.png)
![ButtonPress](https://user-images.githubusercontent.com/962340/82237376-85628780-9903-11ea-9098-728b6247810d.png)

It is impossible to make a 32x32 button using these textures with the margins setup correctly!? (need min 10 margin which makes the button bigger then 32x32!?

**Comment by timothyqiu** [comment]
@rakkarage I think you might need to adjust the font size.

The way Godot computes the minimum control size is somehow unique among the game engines / GUI toolkits.

The centeral stretchable area of the texture is only 12x12 if all the margins are 10. But since the default font size is 16, the 12x12 area can't hold any text using the font at that size, the button's minimum size grows, making it larger than the original 32x32 size.

If a smaller font size is used, which is reasonable for a pixel art style scenario, the minimum height will become smaller. In the screenshot below, the font size is set to 8, and the button's minimum size is 29x32:

![Pixel Button](https://user-images.githubusercontent.com/372476/82274861-d8941300-99b3-11ea-8498-78b53188ea50.png)

The example project: [PixelButton.zip](https://github.com/godotengine/godot/files/4647748/PixelButton.zip)

**Comment by rakkarage** [comment]
Thanks. That works great. i can just make a placeholder dynamic font and set size to 1 if i not want the button NOT to stretch

Though I must disagree that the default size is not reasonable? for a pixel art game as I am using those buttons with these pixel fonts using the default settings and works fine just need to make a 

![Screenshot from 2020-05-19 12-15-47](https://user-images.githubusercontent.com/962340/82351550-b954b000-99ca-11ea-80be-878ed241f287.png)

**Comment by Vhati** [comment]
I too was frustrated then surprised that the font height gets included even when text is "".
I'd've expected that to be conditional.

[Excerpt](https://github.com/godotengine/godot/blob/a743bcef4dfaf97c128a245121709f5f23de7a25/scene/gui/button.cpp#L64-L69): button.cpp, get_minimum_size()
````
	Ref<Font> font = get_theme_font(SNAME("font"));
	float font_height = font->get_height(get_theme_font_size(SNAME("font_size")));

	minsize.height = MAX(font_height, minsize.height);

	return get_theme_stylebox(SNAME("normal"))->get_minimum_size() + minsize;
````

.
.

> **rakkarage:** i can just make a placeholder dynamic font and set size to 1

In v4.0.alpha.calinou [e363f404a], I tried leaving the default font and just setting the font size (Control:Theme Overrides/Font Sizes/Font Size).

* Setting to 1 seemed to have no effect on the minimum size.
* Setting to 2 DID reduce the minimum size.
* Setting to 1, after having set it to 2, started working as expected, reducing the minimum size further than 2 did.

Given the inconsistent results, I'd recommend 2 for folks looking for a workaround to get shorter buttons... if they don't mind a couple pixels of unavoidable minimum height. I didn't need a placeholder font.
.
.
Also surprisingly, when expand_icon is true, hseparation (space between text and icon) is included in the draw calculation.

[Excerpt](https://github.com/godotengine/godot/blob/a743bcef4dfaf97c128a245121709f5f23de7a25/scene/gui/button.cpp#L257-L259): button.cpp, _notification(NOTIFICATION_DRAW)
````
if (!_icon.is_null()) {
	// ...
	if (expand_icon) {
		Size2 _size = get_size() - style->get_offset() * 2;
		_size.width -= get_theme_constant(SNAME("hseparation")) + icon_ofs_region;
		// ...
		if (icon_width > _size.width) {
			icon_width = _size.width;
			icon_height = _icon->get_height() * icon_width / _icon->get_width();
		}
````

Setting "Expand Icon" on a button already at minimum size scales down the icon, when I would've expected it to remain unscaled at that size.

The scaling occurs because the default hseparation=2 eats away at the icon's width. It effectively creates a single pixel gap on the left and right of the icon. A gap between the icon and the stylebox's border.

Again, I would not have expected hseparation to apply when text is "".

It's even conditional already: when NO icon is set, hseparation is ignored. Left & right stylebox borders can squeeze together to touch.

As a workaround, I'd recommend setting Control:Theme Overrides/Constants/Hseparation=0.
.
.
**tl;dr:** Using both workarounds (the small font size and zeroed hseparation), folks can make icon-only buttons whose content meets the border, vertically & horizontally at the icon's native size.

**Comment by KoBeWi** [comment]
Maybe clip text should be changed to enum? So that there is none, horizontal (current), vertical and all. With all clip, text size is completely ignored.
